modules/extension/wmts/src/main/java/org/geotools/data/wmts/client/WMTSTile.java

```diff
@@ -0,0 +1,232 @@
+        if(!baseUrl.contains("?")) {
+
+            baseUrl+="?";
+        }
+        HashMap<String, Object> params = new HashMap<>();
+        params.put("service", "WMTS");
+        params.put("version", "1.0.0");
+        params.put("request", "getTile");
+        params.put("layer",  service.getLayerName());
+        params.put("style",  service.getStyleName());
+        params.put("format", service.getFormat());
+        params.put("tilematrixset", service.getTileMatrixSetName());
+//        params.put("TileMatrix", service.getTileMatrixSetName()+":"+tileIdentifier.getZ());
+//        params.put("TileMatrix", tileIdentifier.getZ());
+        params.put("TileMatrix", service.getTileMatrix(tileIdentifier.getZ()).getIdentifier());
+        params.put("TileCol", tileIdentifier.getX());
+        params.put("TileRow", tileIdentifier.getY());
+
+        StringBuilder arguments = new StringBuilder();
+        for(String p:params.keySet()) {
+            try {
+                arguments.append(p).append("=").append(URLEncoder.encode(params.get(p).toString(),"UTF-8"));
+                arguments.append('&');
+            }catch (Exception e) {
+                // TODO: handle exception
+            }
+        }
+        try {
+            return new URL(baseUrl+arguments.toString());
+        } catch (MalformedURLException e) {
+            //I'm pretty sure this never happens!
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Load and cache locally the WMTS tiles
+     */
+    @Override
+    public BufferedImage loadImageTileImage(Tile tile) throws IOException {
+        LOGGER.log(Level.FINE, "Loading tile "+getId()+": " + this.getUrl());
+
+        String tileKey = tile.getUrl().toString();
+
+        if( !(tileImages.peek(tileKey) == null || tileImages.get(tileKey) == null)) {
+            LOGGER.log(Level.FINE, "Tile image already loaded for tile "+getId());
+            return (BufferedImage) tileImages.get(tileKey);
+        } else {
+            LOGGER.log(Level.FINE, "Tile image not yet loaded for tile "+getId());
+            BufferedImage bi = doLoadImageTileImage(tile);
+            tileImages.put(tileKey, bi);
+            return bi;
+        }
+    }
+
+    public BufferedImage doLoadImageTileImage(Tile tile) throws IOException {
+
+        Map<String, String> headers = (Map<String, String>)this.service.getExtrainfo().get(WMTSTileService.EXTRA_HEADERS);
+        InputStream is = null;
+        try {
+            is = setupInputStream(getUrl(), headers);
+            return ImageIO.read(is);
+        } finally {
+            IOUtils.closeQuietly(is);
+        }
+    }
+
+    private InputStream setupInputStream(URL url, Map<String, String> headers) throws IOException
+    {
+        HttpClient client = new HttpClient();
+        String uri = url.toExternalForm();
+        uri = uri.replaceAll("\\{Time\\}", "20140101");
```

Comment:
Whoa, old stubbed value. By the time the uri reaches this stage, all the tokens should already have been replaced since long.